core/server/worker/src/test/java/alluxio/worker/block/UfsIoManagerTest.java

```diff
@@ -0,0 +1,141 @@
+            UnderFileSystemConfiguration.defaults(Configuration.global())),
+        new AlluxioURI(mTestFilePath));
+    mUfsIOManager = spy(new UfsIOManager(client));
+    doAnswer(args -> {
+      if (mPassQuota.get()) {
+        return 0d;
+      }
+      return 3d * Constants.MB;
+    }).when(mUfsIOManager).getUsedThroughput(any(Meter.class));
+    mUfsIOManager.start();
+  }
+
+  @Test
+  public void readFullBlock() throws Exception {
+    CompletableFuture<byte[]> data = mUfsIOManager.read(FIRST_BLOCK_ID, TEST_BLOCK_SIZE,
+        mTestFilePath, 0, TEST_BLOCK_SIZE, false, "test");
+    assertTrue(BufferUtils.equalIncreasingByteArray(0, (int) TEST_BLOCK_SIZE, data.get()));
+  }
+
+  @Test
+  public void readPartialBlock() throws Exception {
+    CompletableFuture<byte[]> data = mUfsIOManager.read(FIRST_BLOCK_ID, TEST_BLOCK_SIZE,
+        mTestFilePath, 0, TEST_BLOCK_SIZE - 1, false, "test");
+    assertTrue(BufferUtils.equalIncreasingByteArray(0, (int) TEST_BLOCK_SIZE - 1, data.get()));
+  }
+
+  @Test
+  public void readSecondBlock() throws Exception {
+    CompletableFuture<byte[]> data =
+        mUfsIOManager.read(1, TEST_BLOCK_SIZE, mTestFilePath, 0, TEST_BLOCK_SIZE, false, "test");
+    assertTrue(BufferUtils.equalIncreasingByteArray((int) TEST_BLOCK_SIZE, (int) TEST_BLOCK_SIZE,
+        data.get()));
+  }
+
+  @Test
+  public void offset() throws Exception {
+    mUfsIOManager.start();
+    CompletableFuture<byte[]> data = mUfsIOManager.read(FIRST_BLOCK_ID, TEST_BLOCK_SIZE,
+        mTestFilePath, 2, TEST_BLOCK_SIZE - 2, false, "test");
+    assertTrue(BufferUtils.equalIncreasingByteArray(2, (int) TEST_BLOCK_SIZE - 2, data.get()));
+  }
+
+  @Test
+  public void readOverlap() throws Exception {
+    CompletableFuture<byte[]> data = mUfsIOManager.read(FIRST_BLOCK_ID, TEST_BLOCK_SIZE,
+        mTestFilePath, 2, TEST_BLOCK_SIZE - 2, false, "test");
+    CompletableFuture<byte[]> data2 =
+        mUfsIOManager.read(1, TEST_BLOCK_SIZE, mTestFilePath, 0, TEST_BLOCK_SIZE, false, "test");
+    CompletableFuture<byte[]> data3 = mUfsIOManager.read(1, TEST_BLOCK_SIZE, mTestFilePath, 2,
+        TEST_BLOCK_SIZE - 2, false, "test");
+
+    assertTrue(BufferUtils.equalIncreasingByteArray(2, (int) TEST_BLOCK_SIZE - 2, data.get()));
+    assertTrue(BufferUtils.equalIncreasingByteArray((int) TEST_BLOCK_SIZE, (int) TEST_BLOCK_SIZE,
+        data2.get()));
+    assertTrue(BufferUtils.equalIncreasingByteArray((int) TEST_BLOCK_SIZE + 2,
+        (int) TEST_BLOCK_SIZE - 2, data3.get()));
+  }
+
+  @Test
+  public void readExceedQuota() throws Exception {
+    mUfsIOManager.setQuota("quotaTest", 2); // magic number for proper test time
+    CompletableFuture<byte[]> data = mUfsIOManager.read(FIRST_BLOCK_ID, TEST_BLOCK_SIZE,
+        mTestFilePath, 2, TEST_BLOCK_SIZE - 2, false, "quotaTest");
+    // sleep to make sure future is not done because of quota instead of get too soon
+    Thread.sleep(3000);
```

Comment:
This is not necessarily reliable. A more reliable way to test this is to put the logic into a separate function like `getNext` with @VisibleForTesting tag, and have 2 read tasks in the queue with different tag and check that it skips the one that's already exceed bandwidth.

Reply:
things get complicated if we try getNext, basically, you need to have a plausible executor so we can pause the scheduling in the beginning and manually `getNext`. Or we can shut down the executor at the very last test(so it doesn't affect the other test) but we need to introduce sequence order for this test.